src/AESmodes.cpp

```diff
@@ -5,268 +5,811 @@
+        output.clear();
+        return false;
     }
-  }
-
-  printEncryptOutput(output);
+    return true;
 }
 
 
 /**
   Inverse cipher with ECB mode
+  Guaranteed no exceptions by:
+    handling all exceptions per ERR51-CPP
+        Related: Honoring exception specifications, all exceptions will be caught per ERR55-CPP
+    not throwing exceptions across execution boundaries (library to application) per ERR59-CPP
+    Guaranteeing Strong exception safety per ERR56-CPP
+        Program state will not be modified
+            Input and key are constant and output vector is cleared when catching an exception
   @param input: vector of hex values representing (padded) ciphertext
   @param output: vector of hex values representing plaintext (without padding)
   @param key: vector of hex values representing key to use
-  @return none
+  @return True on success
 */
-void decrypt_ecb(std::vector<unsigned char>& input, std::vector<unsigned char>& output, const std::vector<unsigned char>& key) {
-  const std::size_t inputSize = input.size(); // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
-
-  // Loop over number of blocks
-  for (std::size_t i = 0; i < inputSize / 16; i++) { // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
-
-    // Loop over block size and fill each block
-    std::array<unsigned char, 16> block;
-    for (std::size_t j = 0; j < 16; j++) { // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
-      block[j] = input[j+(i*16)]; // Secure coding: OOP57-CPP. Prefer special member functions and overloaded operators to C Standard Library functions
+bool decrypt_ecb(const std::vector<unsigned char> &input, std::vector<unsigned char> &output,
+                 const std::vector<unsigned char> &key) noexcept(true) {
+    try {
+        const std::size_t inputSize = input.size();
+
+        // Loop over number of blocks
+        for (std::size_t i = 0; i < inputSize / NUM_BYTES; i++) {
+
+            // Loop over block size and fill each block
+            std::array<unsigned char, NUM_BYTES> block{0};
+            for (std::size_t j = 0; j < NUM_BYTES; j++) {
+                // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
+                block.at(j) = input.at(j + (i * NUM_BYTES));
+            }
+
+            // Decrypt each block
+            std::array<unsigned char, NUM_BYTES> outputPadded{0};
+            decrypt(block, outputPadded, key);
+
+            // Copy decrypted block to the output
+            for (std::size_t j = 0; j < NUM_BYTES; j++) {
+                // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
+                output.push_back(outputPadded.at(j));
+            }
+        }
+
+        if (!remove_padding(output)) {
```

Comment:
Wouldn't printing this allow a padding oracle attack? `remove_padding` returns `false` at the first incorrect byte, so the error message is also printed at the first incorrect byte.

Reply:
Never mind, this is fine since the error message for padding and for a general decryption error is the same.